File: source/c/edit_distance.c
// gcc -o edit_distance edit_distance.c
#include <stdio.h>
#include <string.h>

#define MAXN 100

int dp[MAXN+1][MAXN+1];
int min3(int a, int b, int c) { return a < b ? (a < c ? a : c) : (b < c ? b : c); }

int main(void) {
    const char *a = "kitten";
    const char *b = "sitting";
    int m = strlen(a), n = strlen(b);

    for (int i = 0; i <= m; i++) dp[i][0] = i;
    for (int j = 0; j <= n; j++) dp[0][j] = j;

    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            if (a[i-1] == b[j-1]) dp[i][j] = dp[i-1][j-1];
            else dp[i][j] = min3(dp[i][j-1]+1, dp[i-1][j]+1, dp[i-1][j-1]+1);
        }
    }

    printf("edit distance: %d\n", dp[m][n]);
    return 0;
}

File: source/rust/edit_distance.rs
fn edit_distance(a: &str, b: &str) -> usize {
    let a: Vec<char> = a.chars().collect();
    let b: Vec<char> = b.chars().collect();
    let m = a.len();
    let n = b.len();
    let mut dp = vec![vec![0usize; n + 1]; m + 1];
    for i in 0..=m { dp[i][0] = i; }
    for j in 0..=n { dp[0][j] = j; }
    for i in 1..=m {
        for j in 1..=n {
            if a[i-1] == b[j-1] {
                dp[i][j] = dp[i-1][j-1];
            } else {
                dp[i][j] = 1 + dp[i][j-1].min(dp[i-1][j]).min(dp[i-1][j-1]);
            }
        }
    }
    dp[m][n]
}

fn main() {
    println!("edit distance: {}", edit_distance("kitten", "sitting"));
}
